File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        
        stack = [] 
        result = [0] * len(temperatures)
        for i, t in enumerate(temperatures):
            if not stack or stack[-1][1] >= t: 
                stack.append((i, t))
            else:
                while stack and stack[-1][1] < t: 
                    index, _ = stack.pop()
                    result[index] = i - index
                stack.append((i, t))
        return result

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<pair<int, int>> temp_and_idx;
        vector<int> result_temperatures(temperatures.size());

        for (int i = 0; i < temperatures.size(); ++i){
            int todays_temp = temperatures[i];
            
            while (!temp_and_idx.empty() && temp_and_idx.top().first < todays_temp){
                int colder_index = temp_and_idx.top().second;
                result_temperatures[colder_index] = i - colder_index;
                temp_and_idx.pop();
            }
            temp_and_idx.push({todays_temp, i});
        }
        
        // while (!temp_and_idx.empty()){
        //     int index = temp_and_idx.pop().second;
        // }

        return result_temperatures;
    }
};